mobile-widgets/uploadDiveLogsDE.cpp

```diff
@@ -0,0 +1,296 @@
+		xmlDoc *transformed;
+		struct zip_source *s;
+		struct membuffer mb = {};
+		
+		/*
+		 * Get the i'th dive in XML format so we can process it.
+		 * We need to save to a file before we can reload it back into memory...
+		 */
+		if (selected && !dive->selected)
+			continue;
+
+		/* make sure the buffer is empty and add the dive */
+		struct dive_site *ds = dive->dive_site;
+		mb.len = 0;
+		if (ds) {
+			put_format(&mb, "<divelog><divesites><site uuid='%8x' name='", ds->uuid);
+			put_quoted(&mb, ds->name, 1, 0);
+			put_format(&mb, "'");
+			put_location(&mb, &ds->location, " gps='", "'");
+			put_format(&mb, ">\n");
+			if (ds->taxonomy.nr) {
+				for (int j = 0; j < ds->taxonomy.nr; j++) {
+					struct taxonomy *t = &ds->taxonomy.category[j];
+					if (t->category != TC_NONE && t->category == prefs.geocoding.category[j] && t->value) {
+						put_format(&mb, "  <geo cat='%d'", t->category);
+						put_format(&mb, " origin='%d' value='", t->origin);
+						put_quoted(&mb, t->value, 1, 0);
+						put_format(&mb, "'/>\n");
+					}
+				}
+			}
+			put_format(&mb, "</site>\n</divesites>\n");
+		}
+		save_one_dive_to_mb(&mb, dive, false);
+
+		if (ds) {
+			put_format(&mb, "</divelog>\n");
+		}
+		membuf = mb_cstring(&mb);
+		streamsize = strlen(membuf);
+		/*
+		 * Parse the memory buffer into XML document and
+		 * transform it to divelogs.de format, finally dumping
+		 * the XML into a character buffer.
+		 */
+		xmlDoc *doc = xmlReadMemory(membuf, streamsize, "divelog", NULL, 0);
+		if (!doc) {
+			qWarning() << errPrefix << "could not parse back into memory the XML file we've just created!";
+			report_error(tr("internal error").toUtf8());
+			zip_close(zip);
+			QFile::remove(filename);
+			xsltFreeStylesheet(xslt);
+			return false;
+		}
+		free_buffer(&mb);
+		
+		transformed = xsltApplyStylesheet(xslt, doc, NULL);
+		if (!transformed) {
+			qWarning() << errPrefix << "XSLT transform failed for dive: " << i;
+			report_error(tr("Conversion of dive %1 to divelogs.de format failed").arg(i).toUtf8());
+			continue;
+		}
+		xmlDocDumpMemory(transformed, (xmlChar **)&membuf, &streamsize);
+		xmlFreeDoc(doc);
+		xmlFreeDoc(transformed);
+		
+		/*
+		 * Save the XML document into a zip file.
+		 */
+		snprintf(xmlfilename, PATH_MAX, "%d.xml", i + 1);
+		s = zip_source_buffer(zip, membuf, streamsize, 1);
+		if (s) {
+			int64_t ret = zip_add(zip, xmlfilename, s);
+			if (ret == -1) {
+				qDebug() << errPrefix << "failed to include dive:" << i;
+			}
+		}
+	}
+	xsltFreeStylesheet(xslt);
+	if (zip_close(zip)) {
+		int ze, se;
+#if LIBZIP_VERSION_MAJOR >= 1
+		zip_error_t *error = zip_get_error(zip);
+		ze = zip_error_code_zip(error);
+		se = zip_error_code_system(error);
+#else
+		zip_error_get(zip, &ze, &se);
+#endif
+		report_error(qPrintable(tr("error writing zip file: %s zip error %d system error %d - %s")),
+					 qPrintable(QDir::toNativeSeparators(filename)), ze, se, zip_strerror(zip));
+		return false;
+	}
+	return true;
+}
```

Comment:
yeah, having compared the complete function the differences feel very very subtle to me - to the point where I'm not quite understanding WHY they are different.

Reply:
I cleaned some return parts, that is all. This is ready as a shared function.